src/compiler/utilities.ts

```diff
@@ -1132,49 +1132,81 @@ namespace ts {
             (<JSDocFunctionType>node).parameters[0].type.kind === SyntaxKind.JSDocConstructorType;
     }
 
-    function getJSDocTag(node: Node, kind: SyntaxKind): JSDocTag {
-        if (node && node.jsDocComment) {
-            for (let tag of node.jsDocComment.tags) {
-                if (tag.kind === kind) {
-                    return tag;
-                }
+    function getJSDocTag(node: Node, kind: SyntaxKind, checkParentVariableStatement: boolean): JSDocTag {
+        if (!node) {
+            return undefined;
+        }
+        
+        const jsDocComment = getJSDocComment(node, checkParentVariableStatement);
+        if (!jsDocComment) {
+            return undefined;
+        }
+        
+        for (let tag of jsDocComment.tags) {
+            if (tag.kind === kind) {
+                return tag;
             }
         }
     }
 
+    function getJSDocComment(node: Node, checkParentVariableStatement: boolean): JSDocComment {
+        if (node.jsDocComment) {
+            return node.jsDocComment;
+        }
+        // Try to recognize this pattern when node is initializer of variable declaration and JSDoc comments are on containing variable statement. 
+        // /** 
+        //   * @param {number} name
+        //   * @returns {number} 
+        //   */
+        // var x = function(name) { return name.length; }
+        if (checkParentVariableStatement) {
+            const isInitializerOfVariableDeclarationInStatement = 
+                node.parent.kind === SyntaxKind.VariableDeclaration && 
+                (<VariableDeclaration>node.parent).initializer === node &&
+                node.parent.parent.parent.kind === SyntaxKind.VariableStatement

```

Comment:
I've considered this as an option and decided not to do this since in JavaScript our goal is to suggest not restrict and it does not really buy us anything. Completion list in JavaScript are populated with
- high confidence items (information that we fish out from type information)
- low confidence items (all identifiers that exist in source files).

In your example in completion list for `name` we'll have some items from the first group and tons of items from the second. For `x` we'll have no high confidence items and still tons of low confidence ones. If we add this restriction then we'll still see the same items in completion lists for both `name` and `x` except that now all of them will have low confidence.